### ISSUE TITLE
ASGI Extension support

### ISSUE THREAD

**Issue by synodriver** [emmett-framework/granian#78]
Great project！Just out of curious, is there any [ASGI Extension support](https://asgi.readthedocs.io/en/latest/extensions.html) in plan?

**Comment by gi0baro** [comment]
@synodriver thank you!

Regarding ASGI extensions, at the moment there are no plans; at least not within the next milestones.
Do you have any specific need/use-case?

**Comment by synodriver** [comment]
Some asgi frameworks use zerocopy extension to serve static files, which can significantly improve the performance, and in some cases the server need to talk to a grpc client directly, which need the trailer header extension. I just forked hypercorn to add [all extension support](https://github.com/nonebot/nonecorn),  this may help implementing these extensions for the rust server.

**Comment by gi0baro** [comment]
So, regarding the two explicit extensions you mentioned:

- due to Granian architecture the zerocopy extension won't be feasible. Since it is based on file-descriptors, I'm 99% sure it won't be possible to make that descriptor available in a non-blocking and safe fashion to the tokio runtime granian uses behind the scenes
- regarding trailers, there might be a chance to make 'em work revising the actual ASGI protocol implementation, but since hyper doesn't support them on HTTP/1.1 (https://github.com/hyperium/hyper/issues/2719), they would work only on http2

**Comment by synodriver** [comment]
Thanks for letting me know that since I'm not very familiar with rust. Looks like the asgi spec itself heavily relies on python asyncio implementation, making  other languages very hard to implement.

**Comment by gi0baro** [comment]
> Looks like the asgi spec itself heavily relies on python asyncio implementation, making other languages very hard to implement.

Yep. It's even in the first paragraph of RSGI rationale (https://github.com/emmett-framework/granian/blob/master/docs/spec/RSGI.md#rationale) ;)

**Comment by synodriver** [comment]
So regardless thoses extension which is difficult to support, what about other extensions mentioned in asgi spec? Are they possible for the rust server？BTW the easiest extension to implement I think should be [lifespan.state](https://github.com/django/asgiref/pull/354), just pass an empty dict to ```scope``` is ok, and [starlette](https://github.com/encode/starlette/pull/2060) have already take use of it.

**Comment by gi0baro** [comment]
@synodriver sorry for the late reply, I was in PyConIT this weekend and a lot was happening!

I agree `lifespan.state` would be the easiest and quicker to implement!
Regarding all the others, maybe we can just put together the complete list and for each of them I can tell if feasible or not :)

**Comment by synodriver** [comment]
I agree too. Actually I have send a pr to nginx unit to add ```lifespan.state support```. It was a C project which I can handle, but unfortunately I'm not familiar with rust =(

**Comment by gi0baro** [comment]
This is now tracked in #93 and #94
